cpp/src/arrow/compute/operation.h

```diff
@@ -0,0 +1,53 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <vector>
+
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class Status;
+
+namespace compute {
+
+class Expr;
+
+/// \brief An operation is a node in a computation graph, taking input data
+/// expression dependencies and emitting an output expression
+class ARROW_EXPORT Operation : public std::enable_shared_from_this<Operation> {
+ public:
+  virtual ~Operation() = default;
+
+  /// \brief Check input expression arguments and output the type of resulting
+  /// expression that this operation produces. If the input arguments are
+  /// invalid, error Status is returned
+  /// \param[out] out the returned well-typed expression
+  /// \return success or failure
+  virtual Status ToExpr(std::shared_ptr<Expr>* out) const = 0;
```

Comment:
I know we use shared_ptr a lot in the code base, but I'm curious if this signature is just habit or you've though through the ownership semantics (i.e. can it be unique_ptr?)

Reply:
I think what @emkornfield is trying to say is: are we using shared_ptr as a way to avoid thinking about lifetime ownership and use it as a poor's man garbage collector. For example, we could get away if there was a parent object owning all the exprs/operations, e.g. 

```
class Graph {
   std::vector<std::shared_ptr<Operation>> ops;
   std::vector<std::shared_ptr<Expr>> exprs;
}
```

With this approach, each expr/ops can use plain pointers and still preserve memory safety. The downside is that the Graph becomes the object that needs to be shared.

Reply:
I don't like that very much. I do not intend these objects as single use, so I think that `shared_ptr` is more appropriate to steward object lifetime

Reply:
@fsaintjacques summarized the point I was trying to make.  I'm ok with this approach as it seems to make sense.  I just wanted to make sure we weren't cargo culting based on existing design.